### PR TITLE
Empty struct init alternate

### PR DESCRIPTION
**Pull request by WalterSmuts**
Fixes: https://github.com/hylo-lang/hylo/issues/997

Alternative approach to https://github.com/hylo-lang/hylo/pull/1014

### CODE REVIEW

File: Sources/IR/Emitter.swift [hunk 1]
```diff
@@ -316,6 +316,15 @@ struct Emitter {
...
     let bodyFrame = Frame(locals: locals)
     let returnSite = pushing(bodyFrame, { $0.lowerStatements($0.ast[d].body!, expecting: .void) })
+
+    let receiverType: AnyType = program[d].receiver.type
+    let layout = AbstractTypeLayout(of: receiverType, definedIn: program)
+    let receiver = entry.parameter(0)
+
+    // If the object is empty, simply mark it initialized.
+    if layout.properties.isEmpty {
+      insert(module.makeMarkState(receiver, initialized: true, at: ast[d].site))
```
Comment (kyouko-taiga) [inline Sources/IR/Emitter.swift:326]:
```suggestion
    // If the object is empty, simply mark it initialized.
    if layout.properties.isEmpty {
      let receiver = entry.parameter(0)
      insert(module.makeMarkState(receiver, initialized: true, at: ast[d].site))
```

File: Sources/IR/Emitter.swift [hunk 2]
```diff
@@ -316,6 +316,15 @@ struct Emitter {
     insertionPoint = .end(of: entry)
     let bodyFrame = Frame(locals: locals)
     let returnSite = pushing(bodyFrame, { $0.lowerStatements($0.ast[d].body!, expecting: .void) })
+
+    let receiverType: AnyType = program[d].receiver.type
+    let layout = AbstractTypeLayout(of: receiverType, definedIn: program)
```
Comment (kyouko-taiga) [inline Sources/IR/Emitter.swift:321]:
```suggestion
    let layout = AbstractTypeLayout(of: program[d].receiver.type, definedIn: program)
```

**Review by kyouko-taiga — APPROVED** [review]
I like this approach much better.

File: Sources/IR/Emitter.swift (hunk 1 above)
Comment (WalterSmuts) [reply Sources/IR/Emitter.swift:326]: Good suggestion. Have included it in my most recent `git push --force`

File: Sources/IR/Emitter.swift (hunk 2 above)
Comment (WalterSmuts) [reply Sources/IR/Emitter.swift:321]: Good suggestion. Have included it in my most recent `git push --force`

**Comment by WalterSmuts** [comment]
> I like this approach much better.

Agreed. There is also a couple of other situations where this approach is already in use:
https://github.com/hylo-lang/hylo/blob/main/Sources/IR/Emitter.swift#L703-L708
https://github.com/hylo-lang/hylo/blob/main/Sources/IR/Emitter.swift#L725-L730
https://github.com/hylo-lang/hylo/blob/main/Sources/IR/Emitter.swift#L2381-L2385
https://github.com/hylo-lang/hylo/blob/main/Sources/IR/Emitter.swift#L2402-L2406

Would be inconsistent not to use the same approach for all these cases.

**Pull request merged**